File: services/user-service/src/test/java/com/visor/school/userservice/integration/UserRegistrationIntegrationTest.java
package com.visor.school.userservice.integration;

import java.util.HashSet;
import java.util.Optional;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

import com.visor.school.userservice.model.User;
import com.visor.school.userservice.model.UserRole;
import com.visor.school.userservice.repository.UserRepository;
import com.visor.school.userservice.service.UserService;

/**
 * Integration test for user registration flow
 * Tests: Keycloak user creation → User entity creation
 * 
 * Note: This test requires Keycloak to be running or Testcontainers
 * TODO: Add Testcontainers support for Keycloak
 */
@SpringBootTest
@ActiveProfiles("test")
@Transactional
@Disabled("Requires Keycloak instance or Testcontainers setup")
class UserRegistrationIntegrationTest {

    @Autowired
    private UserService userService;

    @Autowired
    private UserRepository userRepository;

    @Test
    void shouldCompleteUserRegistrationFlowWithKeycloakIntegration() {
        // Given
        String email = "integration-test-" + System.currentTimeMillis() + "@example.com";
        String firstName = "Integration";
        String lastName = "Test";
        String password = "TestPassword123!";

        Set<UserRole> roles = new HashSet<>();
        roles.add(UserRole.STUDENT);

        // When
        User user = userService.createUser(
            email,
            firstName,
            lastName,
            roles,
            password,
            null
        );

        // Then
        assertNotNull(user);
        assertNotNull(user.getKeycloakId());
        assertFalse(user.getKeycloakId().isBlank());
        assertEquals(email, user.getEmail());
        assertEquals(firstName, user.getFirstName());
        assertEquals(lastName, user.getLastName());
        assertTrue(user.getRoles().contains(UserRole.STUDENT));

        // Verify user is persisted
        Optional<User> savedUser = userRepository.findById(user.getId());
        assertTrue(savedUser.isPresent());
        assertEquals(email, savedUser.get().getEmail());
    }

    @Test
    void shouldFailRegistrationWhenEmailAlreadyExists() {
        // Given
        String email = "duplicate-test-" + System.currentTimeMillis() + "@example.com";
        
        Set<UserRole> roles = new HashSet<>();
        roles.add(UserRole.TEACHER);
        
        userService.createUser(
            email,
            "First",
            "User",
            roles,
            "Password123!",
            null
        );

        // When & Then
        assertThrows(IllegalArgumentException.class, () -> {
            userService.createUser(
                email,
                "Second",
                "User",
                roles,
                "Password123!",
                null
            );
        });
    }
}

File: services/user-service/src/test/kotlin/com/visor/school/userservice/integration/UserRegistrationIntegrationTest.kt
package com.visor.school.userservice.integration

import com.visor.school.userservice.model.UserRole
import com.visor.school.userservice.repository.UserRepository
import com.visor.school.userservice.service.UserService
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional

/**
 * Integration test for user registration flow
 * Tests: Keycloak user creation → User entity creation
 * 
 * Note: This test requires Keycloak to be running or mocked
 */
@SpringBootTest
@ActiveProfiles("test")
@Transactional
class UserRegistrationIntegrationTest @Autowired constructor(
    private val userService: UserService,
    private val userRepository: UserRepository
) {

    @Test
    fun `should complete user registration flow with Keycloak integration`() {
        // Given
        val email = "integration-test-${System.currentTimeMillis()}@example.com"
        val firstName = "Integration"
        val lastName = "Test"
        val password = "TestPassword123!"

        // When
        val user = userService.createUser(
            email = email,
            firstName = firstName,
            lastName = lastName,
            role = UserRole.STUDENT,
            password = password
        )

        // Then
        assertNotNull(user)
        assertNotNull(user.keycloakId)
        assertFalse(user.keycloakId.isBlank())
        assertEquals(email, user.email)
        assertEquals(firstName, user.firstName)
        assertEquals(lastName, user.lastName)
        assertEquals(UserRole.STUDENT, user.role)

        // Verify user is persisted
        val savedUser = userRepository.findById(user.id)
        assertTrue(savedUser.isPresent)
        assertEquals(email, savedUser.get().email)
    }

    @Test
    fun `should fail registration when email already exists`() {
        // Given
        val email = "duplicate-test-${System.currentTimeMillis()}@example.com"
        userService.createUser(
            email = email,
            firstName = "First",
            lastName = "User",
            role = UserRole.TEACHER,
            password = "Password123!"
        )

        // When & Then
        assertThrows<IllegalArgumentException> {
            userService.createUser(
                email = email,
                firstName = "Second",
                lastName = "User",
                role = UserRole.TEACHER,
                password = "Password123!"
            )
        }
    }
}

